### PR TITLE
#181339532-ResetPassword-via-email

### PR DESCRIPTION
**Pull request by niyonkurueric**
### What does this PR do?
his PR implements Forgot-password and reset the password via Email
### Description of Task to be completed?

- Allow a user to forget password
- user forgot the password  if he is already in the database
- Allow a user to Reset Password
### Relevant pivotal tracker stories?
[#181339532](https://www.pivotaltracker.com/story/show/181339532)
### How can you manually test this
- git clone repository
- run you server with `yarn run dev` (remember to setup you dev database)
- for testing run your server by `yarn run test` (remember to setup you test database)
- testing these routes in postman: `/api/v1/users/forgot-password`
- test this route for authentication: `/api/v1/users/reset-password`
### How can you  test in online 
[resetPassword](https://staging-bare-feature-18-fvn8zj.herokuapp.com/api-docs/)
### screenshot
Screenshot from 2022-03-16 20-54-53

### CODE REVIEW

**Review by Eli250 — APPROVED** [review]

**Review by blandri — APPROVED** [review]

File: src/controllers/user.controller.js [hunk 1]
```diff
@@ -1,53 +1,101 @@
...
-      data.password = hashPassword(data.password);
-      data.role_id = 4;
-      const newUser = await this.userService.createUser(data, res);
+        if (exist.email) {
+            const tokenid = generateToken({ id: exist.id }, '10m')
+            const message = {
+                to: exist.email,
+                from: {
+                    name: "cabal",
+                    email: 'cabalbarefoot@gmail.com',
```
Comment (ncutixavier) [inline src/controllers/user.controller.js]: This info needs to be in .env file

File: src/controllers/user.controller.js [hunk 2]
```diff
@@ -1,53 +1,101 @@
...
-        .json({ message: 'user registered successfully', token });
-    } catch (error) {
-      return res.status(500).json({
-        message: 'Error occured while creating a user',
-        error:
-          (error.message && error.message.replace(/['"`]+/gi, '')) ||
-          'user not created'
-      });
+        } else {
+            res.status(403).json({ status: 403, message: 'Email not exist' });
```
Comment (ncutixavier) [inline src/controllers/user.controller.js]: - Change status code for not found code
- Change message to "Email has not found"

File: src/controllers/user.controller.js [hunk 3]
```diff
@@ -1,53 +1,101 @@
...
+        try {
+            const { password } = req.body;
+            const token = req.params.token;
+            const userInfo = jwt.verify(token, process.env.SECRETE);
+            const userId = userInfo.id;
+            const newPassword = await bcrypt.hash(password, 10);
+            User.update({ password: newPassword }, { where: { id: userId } })
+            return res.status(200).send({ message: "reset well continue to login" });
+        } catch (error) {
+            return res.status(500).send({ message: error.message });
```
Comment (ncutixavier) [inline src/controllers/user.controller.js:100]: Provide appropriate error for the user

File: src/controllers/user.controller.js [hunk 4]
```diff
@@ -1,53 +1,101 @@
...
-      });
+        } else {
+            res.status(403).json({ status: 403, message: 'Email not exist' });
+        }
     }
-  }
+    static async reset(req, res) {
+        try {
+            const { password } = req.body;
+            const token = req.params.token;
```
Comment (ncutixavier) [inline src/controllers/user.controller.js]: implement object destructuring

File: src/controllers/user.controller.js (hunk 3 above)
Comment (ncutixavier) [reply src/controllers/user.controller.js:100]: Also, refactor your code to remain with few variables

File: src/documentation/forgot.doc.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import responses from './response';
+
+
+const forgot = {
+    '/users/forgot': {
```
Comment (ncutixavier) [inline src/documentation/forgot.doc.js]: change route to " /users/forgot-password", and implement the same on reset password route

File: src/documentation/forgot.doc.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+                    name: 'name',
+                    required: true,
+                    schema: {
+                        example: {
+                            password: 'password',
+                        },
+                    },
+                },
+
+            ],
```
Comment (ncutixavier) [inline src/documentation/forgot.doc.js:42]: Format your code

File: src/email/index.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+                            <tr>
+                                <td style="height:80px;">&nbsp;</td>
+                            </tr>
+                    </table>
+                </td>
+                </tr>
+        </table>
+    </body>
+    
+    </html>`;
```
Comment (ncutixavier) [inline src/email/index.js]: Create files in the `utils` folder and add these codes.

File: tests/password-reset.test.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+            .send({ password: "4assword" })
+            .end((err, res) => {
+                expect(res).to.have.status([400]);
+                done();
+            });
+    });
+    it('should reset not  password', (done) => {
+        chai
+            .request(app)
+            .patch(`/api/v1/users/reset/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6NCwiaWF0IjoxNjQ3MTUzNTY4LCJleHAiOjE2NDcxNTQxNjh9.k8jHrhjRJXymilhpotQTVjhxhF4_4jziU2w1tHTh3WA`)
```
Comment (ncutixavier) [inline tests/password-reset.test.js]: Reduce the length of this line

**Review by ncutixavier — CHANGES_REQUESTED** [review]

**Comment by ncutixavier** [comment]
image

**Review by ncutixavier — COMMENTED** [review]
- Rephrase the message when reset is done
image

**Review by ncutixavier — APPROVED** [review]

**Review by ncutixavier — APPROVED** [review]

**Pull request merged**